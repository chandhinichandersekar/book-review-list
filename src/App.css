.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 40vmin;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}


#booklist {
  border-collapse: collapse;
  width: 90%;

}

#booklist tr:hover {
  background-color: #C7EAFB;
}

#booklist td, #booklist th {
  border: 1px solid #ddd;
  padding: 8px;
}

#booklist th {
  padding-top: 12px;
  padding-bottom: 12px;
  text-align: center;
  background-color: #056593;
  color: white;
}