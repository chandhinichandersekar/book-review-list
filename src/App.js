import React from 'react';
import './App.css';
const superagent = require('superagent');

async function getBooks(query) {
  return new Promise(resolve => {
    superagent.get('https://find-books-by-title-author-or-isbn-ztnewtvsrs9y.runkit.sh/').query({ ...query }).then(res => {
      resolve(res.body);
    })
  });
}

export class BookRows extends React.Component {
  render() {
    const bookRows = this.props.books.map(book => {
      return (
        <tbody>
        <tr key={book.title}>
        <TdWithStyles>{book.title}</TdWithStyles>
        <TdWithStyles>{book.author}</TdWithStyles>
        <TdWithStyles>{book.average_rating}</TdWithStyles>
        </tr>
        </tbody>
      )
    })
    return bookRows;
  }
}

function TdWithStyles(props) {
  const styles = {
    "border": "1px solid #ddd",
    "padding": "8px",
  };
  return (
    <td style={styles}>{props.children}</td>  
  );
}

function ThWithStyles(props) {
  const styles = {
    "border": "1px solid #ddd",
    "padding": "8px",
    "paddingTop": "12px",
    "paddingBottom": "12px",
    "textAlign": "center",
    "backgroundColor": "#056593",
    color: "white"
  };
  return (
    <th style={styles}>{props.children}</th>  
  );
}

class BookTable extends React.Component {

  render() {
    const styles = {
        "borderCollapse": "collapse",
        width: "90%"
    };
    return (
      <table id="booklist" style={styles}>
        <thead>
          <tr>
            <ThWithStyles>Title</ThWithStyles>
            <ThWithStyles>Author</ThWithStyles>
            <ThWithStyles>Average Rating</ThWithStyles>
          </tr>
        </thead>
          <BookRows books={this.props.books} />
      </table>
    )
  }
}

export class SearchField extends React.Component {

  constructor(props) {
    super(props);
    this.state = { value: '' };
  }

  handleChange(event) {
    this.props.onChange(event.target.value);
  }

  render() {
    return (
      <label>
        Enter text to search:
          <input type="text" value={this.props.value} onChange={this.handleChange.bind(this)} />
      </label>
    )
  }
}

export class Pagination extends React.Component {

  constructor(props) {
    super(props);
    this.state = { value: '' };
  }

  handleChange(event) {
    this.props.onChange(event.target.value);
  }


  render() {
    return (
        <label>
          Pagination Parameters:
          <input type="number" value={this.props.value} onChange={this.handleChange.bind(this)} />
        </label>
    )
  }
}

export class SearchType extends React.Component {

  constructor(props) {
    super(props);
    this.state = { value: '' };
  }

  handleChange(event) {
    this.props.onChange(event.target.value);
  }

  render() {
    return (
        <label>
          Choose a search type parameter:
        <select value={this.props.value} onChange={this.handleChange.bind(this)}>
            <option value="all">all</option>
            <option value="author">author</option>
            <option value="title">title</option>
          </select>
        </label>
    )
  }
}

class App extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      books: [],
      query: {
        text: 'ch',
        page: '',
        search: ''
      }
    };
  }

  async handleSubmit(e) {
    e.preventDefault();
    this.search();
  }

  async search() {
    const books = await getBooks(this.state.query);
    this.setState({
      books
    });
  }

  handleText(text) {
    this.setState({
      query: {
        ...this.state.query,
        text
      }
    });
  }

  handlePage(page) {
    this.setState({
      query: {
        ...this.state.query,
        page
      }
    });
  }

  handleSearch(search) {
    this.setState({
      query: {
        ...this.state.query,
        search
      }
    });
  }


  async componentDidMount() {
    this.search()
  }
  render() {
    // const query = { text: "ch" };
    return (
      <div>
        <form onSubmit={this.handleSubmit.bind(this)}>
          <SearchField onChange={this.handleText.bind(this)} value={this.state.query.text}/>
          <Pagination onChange={this.handlePage.bind(this)} value={this.state.query.page}/>
          <SearchType onChange={this.handleSearch.bind(this)} value={this.state.query.search}/>
          <input type="submit" value="submit"></input>
        </form>
        <BookTable books={this.state.books} />
      </div>
    )
  }

}

export default App;

