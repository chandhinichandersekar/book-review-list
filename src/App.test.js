import React from 'react';
import ReactDOM from 'react-dom';
import App, { BookRows, SearchField, Pagination } from './App';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
configure({ adapter: new Adapter() });
import toJson from 'enzyme-to-json';

const books = [{ "title": "ANIMAL FARM (AB)", "author": "George Orwell", "average_rating": "4.26" }, { "title": "Ardil-22 (Catch-22, # 1)", "author": "Joseph Heller", "average_rating": "3.26" }, { "title": "A Tale of Two Cities", "author": "Charles Dickens", "average_rating": "4.16" }, { "title": "Fields of Elysium (Fields of Elysium, #1)", "author": "A.B. Whelan", "average_rating": "4.26" }, { "title": "14 Days to Die", "author": "A.B. Whelan", "average_rating": "4.26" }, { "title": "Cornelsen Senior English Library - Literatur: Ab 11. Schuljahr - Never let me go: Textband mit Annotationen und Zusatztexten", "author": "Kazuo Ishiguro", "average_rating": "4.26"  }, { "title": "A Leste do Éden (Vol I e II)", "author": "John Steinbeck", "average_rating": "4.26"  }, { "title": "Atlas wirft die Welt ab", "author": "Ayn Rand", "average_rating": "4.26"  }, { "title": "The Way West (The Big Sky, #2)", "author": "A.B. Guthrie Jr.", "average_rating": "4.26"  }, { "title": "Hart van Inkt (Inkt, #1)", "author": "Cornelia Funke", "average_rating": "4.26"  }, { "title": "The Big Sky (The Big Sky, #1)", "author": "A.B. Guthrie Jr." , "average_rating": "4.26" }, { "title": "As Sick as Our Secrets", "author": "A.B. Whelan", "average_rating": "4.26"  }, { "title": "AB de Villiers - The Autobiography", "author": "A.B. de Villiers", "average_rating": "4.26"  }, { "title": "Oliver Twist ( Ab 11 J.).", "author": "Charles Dickens", "average_rating": "4.26"  }, { "title": "Don Quixote [illustrated]", "author": "Miguel de Cervantes Saavedra", "average_rating": "4.26"  }, { "title": "Insel der blauen Delfine. CD. ( Ab 8 J.)", "author": "Scott O'Dell" , "average_rating": "4.26" }, { "title": "A Streetcar Named Desire. Ab 11. Schuljahr", "author": "Tennessee Williams" , "average_rating": "4.26" }, { "title": "A mulher de Jerusalém", "author": "A.B. Yehoshua", "average_rating": "4.26"  }, { "title": "The Lover", "author": "A.B. Yehoshua" , "average_rating": "4.26" }, { "title": "O misterioso caso de Styles", "author": "Agatha Christie", "average_rating": "4.26"  }]

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('when App is rendered with shallow', () => {
  it('should match the snapshot', () => {
    const wrapper = shallow(
      <App />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
})

describe('when BookRows is rendered with shallow', () => {
  describe('before component has mounted', () => {
    it('should match the snapshot', () => {
      const wrapper = shallow(
        <BookRows books={books} />
      );
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});

describe('when SearchField is rendered with shallow', () => {
  const mockHandleSubmit = jest.fn();
  const wrapper = shallow(
    <SearchField handleSubmit={mockHandleSubmit} />
  );
  it('should match the snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  describe('when input is changed', () => {
    it('should update the input value', () => {
      wrapper.find('input').first().simulate('change', {
        target: {
          value: 'typedSearch'
        }
      });
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
  describe('when submitted', () => {
    it('should update the input value', () => {
      wrapper.instance().handleSubmit({
        preventDefault: jest.fn()
      });
      const mockCall = mockHandleSubmit.mock.calls[0];
      expect(mockCall).toMatchObject([
        'typedSearch'
      ]);
    });
  });
});


describe('when Pagination is rendered with shallow', () => {
  const mockHandleSubmit = jest.fn();
  const wrapper = shallow(
    <Pagination handleSubmit={mockHandleSubmit} />
  );
  it('should match the snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  describe('when input is changed', () => {
    it('should update the input value', () => {
      wrapper.find('input').first().simulate('change', {
        target: {
          value: '1'
        }
      });
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
  describe('when submitted', () => {
    it('should update the input value', () => {
      wrapper.instance().handleSubmit({
        preventDefault: jest.fn()
      });
      const mockCall = mockHandleSubmit.mock.calls[0];
      expect(mockCall).toMatchObject([
        '1'
      ]);
    });
  });
});
