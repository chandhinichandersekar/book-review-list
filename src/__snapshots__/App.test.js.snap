// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when App is rendered with shallow should match the snapshot 1`] = `
<div>
  <SearchField
    handleSubmit={[Function]}
  />
  <Pagination
    handleSubmit={[Function]}
  />
  <BookTable
    books={Array []}
  />
</div>
`;

exports[`when BookRows is rendered with shallow before component has mounted should match the snapshot 1`] = `
Array [
  <tr
    key="ANIMAL FARM (AB)"
  >
    <td>
      ANIMAL FARM (AB)
    </td>
    <td>
      George Orwell
    </td>
  </tr>,
  <tr
    key="Ardil-22 (Catch-22, # 1)"
  >
    <td>
      Ardil-22 (Catch-22, # 1)
    </td>
    <td>
      Joseph Heller
    </td>
  </tr>,
  <tr
    key="A Tale of Two Cities"
  >
    <td>
      A Tale of Two Cities
    </td>
    <td>
      Charles Dickens
    </td>
  </tr>,
  <tr
    key="Fields of Elysium (Fields of Elysium, #1)"
  >
    <td>
      Fields of Elysium (Fields of Elysium, #1)
    </td>
    <td>
      A.B. Whelan
    </td>
  </tr>,
  <tr
    key="14 Days to Die"
  >
    <td>
      14 Days to Die
    </td>
    <td>
      A.B. Whelan
    </td>
  </tr>,
  <tr
    key="Cornelsen Senior English Library - Literatur: Ab 11. Schuljahr - Never let me go: Textband mit Annotationen und Zusatztexten"
  >
    <td>
      Cornelsen Senior English Library - Literatur: Ab 11. Schuljahr - Never let me go: Textband mit Annotationen und Zusatztexten
    </td>
    <td>
      Kazuo Ishiguro
    </td>
  </tr>,
  <tr
    key="A Leste do Éden (Vol I e II)"
  >
    <td>
      A Leste do Éden (Vol I e II)
    </td>
    <td>
      John Steinbeck
    </td>
  </tr>,
  <tr
    key="Atlas wirft die Welt ab"
  >
    <td>
      Atlas wirft die Welt ab
    </td>
    <td>
      Ayn Rand
    </td>
  </tr>,
  <tr
    key="The Way West (The Big Sky, #2)"
  >
    <td>
      The Way West (The Big Sky, #2)
    </td>
    <td>
      A.B. Guthrie Jr.
    </td>
  </tr>,
  <tr
    key="Hart van Inkt (Inkt, #1)"
  >
    <td>
      Hart van Inkt (Inkt, #1)
    </td>
    <td>
      Cornelia Funke
    </td>
  </tr>,
  <tr
    key="The Big Sky (The Big Sky, #1)"
  >
    <td>
      The Big Sky (The Big Sky, #1)
    </td>
    <td>
      A.B. Guthrie Jr.
    </td>
  </tr>,
  <tr
    key="As Sick as Our Secrets"
  >
    <td>
      As Sick as Our Secrets
    </td>
    <td>
      A.B. Whelan
    </td>
  </tr>,
  <tr
    key="AB de Villiers - The Autobiography"
  >
    <td>
      AB de Villiers - The Autobiography
    </td>
    <td>
      A.B. de Villiers
    </td>
  </tr>,
  <tr
    key="Oliver Twist ( Ab 11 J.)."
  >
    <td>
      Oliver Twist ( Ab 11 J.).
    </td>
    <td>
      Charles Dickens
    </td>
  </tr>,
  <tr
    key="Don Quixote [illustrated]"
  >
    <td>
      Don Quixote [illustrated]
    </td>
    <td>
      Miguel de Cervantes Saavedra
    </td>
  </tr>,
  <tr
    key="Insel der blauen Delfine. CD. ( Ab 8 J.)"
  >
    <td>
      Insel der blauen Delfine. CD. ( Ab 8 J.)
    </td>
    <td>
      Scott O'Dell
    </td>
  </tr>,
  <tr
    key="A Streetcar Named Desire. Ab 11. Schuljahr"
  >
    <td>
      A Streetcar Named Desire. Ab 11. Schuljahr
    </td>
    <td>
      Tennessee Williams
    </td>
  </tr>,
  <tr
    key="A mulher de Jerusalém"
  >
    <td>
      A mulher de Jerusalém
    </td>
    <td>
      A.B. Yehoshua
    </td>
  </tr>,
  <tr
    key="The Lover"
  >
    <td>
      The Lover
    </td>
    <td>
      A.B. Yehoshua
    </td>
  </tr>,
  <tr
    key="O misterioso caso de Styles"
  >
    <td>
      O misterioso caso de Styles
    </td>
    <td>
      Agatha Christie
    </td>
  </tr>,
]
`;

exports[`when SearchField is rendered with shallow should match the snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <label>
    Search Parameters:
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <button
    type="submit"
    value="Submit"
  />
</form>
`;

exports[`when SearchField is rendered with shallow when input is changed should update the input value 1`] = `
<form
  onSubmit={[Function]}
>
  <label>
    Search Parameters:
    <input
      onChange={[Function]}
      type="text"
      value="typedSearch"
    />
  </label>
  <button
    type="submit"
    value="Submit"
  />
</form>
`;
